api/provider/aws/environment_unlink.go

```diff
@@ -0,0 +1,100 @@
+package aws
+
+import (
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/quintilesims/layer0/common/models"
+)
+
+func (e *EnvironmentProvider) Unlink(req models.DeleteEnvironmentLinkRequest) error {
+	log.Printf("[DEBUG] unlink called")
+	if err := models.EnvironmentLinkRequest(req).Validate(); err != nil {
+		return err
+	}
+
+	fqSourceEnvID := addLayer0Prefix(e.Config.Instance(), req.SourceEnvironmentID)
+	fqDestEnvID := addLayer0Prefix(e.Config.Instance(), req.DestEnvironmentID)
+
+	sourceSecurityGroup, err := readSG(e.AWS.EC2, getEnvironmentSGName(fqSourceEnvID))
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "DoesNotExist" {
+			log.Printf("[WARN] skipping environment unlink since security group '%s' does not exist\n", getEnvironmentSGName(fqSourceEnvID))
+			return nil
+		}
+
+		return err
+	}
+
+	destSecurityGroup, err := readSG(e.AWS.EC2, getEnvironmentSGName(fqDestEnvID))
+	if err != nil {
+		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == "DoesNotExist" {
+			log.Printf("[WARN] skipping environment unlink since security group '%s' does not exist\n", getEnvironmentSGName(fqDestEnvID))
+			return nil
+		}
+
+		return err
+	}
+
+	sourceSecurityGroupID := aws.StringValue(sourceSecurityGroup.GroupId)
+	destSecurityGroupID := aws.StringValue(destSecurityGroup.GroupId)
+
+	if err := e.removeIngressRule(sourceSecurityGroup, destSecurityGroupID); err != nil {
+		return err
+	}
+
+	if err := e.removeIngressRule(destSecurityGroup, sourceSecurityGroupID); err != nil {
+		return err
+	}
+
+	if err := e.deleteLinkTag(req.SourceEnvironmentID, req.DestEnvironmentID); err != nil {
+		return err
+	}
+
+	if err := e.deleteLinkTag(req.DestEnvironmentID, req.SourceEnvironmentID); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *EnvironmentProvider) removeIngressRule(group *ec2.SecurityGroup, groupIDToRemove string) error {
+	for _, permission := range group.IpPermissions {
```

Comment:
this implementation seems heavy-handed. Since we know what the `IpPermission` will look like, can't we just call `RevokeSecurityGroupIngress` and catch idempotent errors? That would also make this easier to test (if we decide to unit test helpers... yet to see i guess)